File: day22_part1_2017.js
const fs = require('fs');

const input = fs.readFileSync('input.txt', 'utf8').split('\n');

const grid = new Map();
let startX, startY;

input.forEach((line, y) => {
    [...line].forEach((c, x) => {
        if (c === '#') {
            grid.set(`${x},${y}`, true);
        }
    });
    startX = Math.floor(line.length / 2);
    startY = Math.floor(y / 2);
});

const dx = [0, 1, 0, -1];
const dy = [-1, 0, 1, 0];

let x = startX;
let y = startY;
let dir = 0;
let infectedCount = 0;

for (let i = 0; i < 10000; i++) {
    const pos = `${x},${y}`;
    if (grid.has(pos)) {
        dir = (dir + 1) % 4;
        grid.delete(pos);
    } else {
        dir = (dir - 1 + 4) % 4;
        grid.set(pos, true);
        infectedCount++;
    }
    x += dx[dir];
    y += dy[dir];
}

console.log(infectedCount);

File: java/day22_part1_2017.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Main {
    static class Position {
        int x, y;
        Position(int x, int y) {
            this.x = x;
            this.y = y;
        }
        @Override
        public int hashCode() {
            return x * 31 + y;
        }
        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            Position p = (Position) obj;
            return x == p.x && y == p.y;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new FileReader("input.txt"));
        Map<Position, Boolean> grid = new HashMap<>();
        int startX = 0, startY = 0;
        int yCoord = 0;
        while (true) {
            String line = br.readLine();
            if (line == null) break;
            for (int x = 0; x < line.length(); x++) {
                if (line.charAt(x) == '#') {
                    grid.put(new Position(x, yCoord), true);
                }
            }
            startX = line.length() / 2;
            startY = yCoord / 2;
            yCoord++;
        }

        int[] dx = {0, 1, 0, -1};
        int[] dy = {-1, 0, 1, 0};
        int xCoord = startX, yCoordCurr = startY, dir = 0;
        int infectedCount = 0;

        for (int i = 0; i < 10000; i++) {
            Position pos = new Position(xCoord, yCoordCurr);
            if (grid.containsKey(pos)) {
                dir = (dir + 1) % 4;
                grid.remove(pos);
            } else {
                dir = (dir - 1 + 4) % 4;
                grid.put(pos, true);
                infectedCount++;
            }
            xCoord += dx[dir];
            yCoordCurr += dy[dir];
        }

        System.out.println(infectedCount);
    }
}